src/js/game/LevelMVC/LevelPlane.js

```diff
@@ -325,10 +326,168 @@ module.exports = class LevelPlane extends Array {
+    let neighborPosition = this.getOrthogonalPositions(position);
+
+    switch (this[this.coordinatesToIndex(position)].blockType) {
+      case "pistonUp": {
+        if (neighbors.north.block.blockType !== "" && neighbors.north.block.blockType !== "pistonArm" ) {
+          let offset = [0,-1];
+          let blocksPositions = this.getBlocksToPush(neighborPosition[0], offset[0], offset[1]);
+          this.pushBlocks(blocksPositions, offset[0], offset[1]);
+        } else if (neighbors.north.block.blockType === "") {
+          let armBlock = new LevelBlock("pistonArm");
+          this.setBlockAt(neighborPosition[0], armBlock);
+        }
+        break;
+      }
+      case "pistonDown": {
+        if (neighbors.south.block.blockType !== "" && neighbors.south.block.blockType !== "pistonArm") {
+          let offset = [0,1];
+          let blocksPositions = this.getBlocksToPush(neighborPosition[1], offset[0], offset[1]);
+          this.pushBlocks(blocksPositions, offset[0], offset[1]);
+        } else if (neighbors.south.block.blockType === "") {
+          let armBlock = new LevelBlock("pistonArm");
+          this.setBlockAt(neighborPosition[1], armBlock);
+        }
+        break;
+      }
+      case "pistonRight": {
+        if (neighbors.east.block.blockType !== "" && neighbors.east.block.blockType !== "pistonArm") {
+          let offset = [1,0];
+          let blocksPositions = this.getBlocksToPush(neighborPosition[2], offset[0], offset[1]);
+          this.pushBlocks(blocksPositions, offset[0], offset[1]);
+        } else if (neighbors.east.block.blockType === "") {
+          let armBlock = new LevelBlock("pistonArm");
+          this.setBlockAt(neighborPosition[2], armBlock);
+        }
+        break;
+      }
+      case "pistonLeft": {
+        if (neighbors.west.block.blockType !== "" && neighbors.west.block.blockType !== "pistonArm") {
+          let offset = [-1,0];
+          let blocksPositions = this.getBlocksToPush(neighborPosition[3], offset[0], offset[1]);
+          this.pushBlocks(blocksPositions, offset[0], offset[1]);
+        } else if (neighbors.west.block.blockType === "") {
+          let armBlock = new LevelBlock("pistonArm");
+          this.setBlockAt(neighborPosition[3], armBlock);
+        }
+        break;
+      }
+    }
+  }
+
+  /**
+  * Deactivates a piston at a given position by determining what the arm orientation is.
+  */
+  deactivatePiston(position) {
+    let neighborPosition = this.getOrthogonalPositions(position);
+    switch (this[this.coordinatesToIndex(position)].blockType) {
+      case "pistonUp": {
+        this.retractArm(neighborPosition[0]);
+        break;
+      }
+      case "pistonDown": {
+        this.retractArm(neighborPosition[1]);
+        break;
+      }
+      case "pistonRight": {
+        this.retractArm(neighborPosition[2]);
+        break;
+      }
+      case "pistonLeft": {
+        this.retractArm(neighborPosition[3]);
```

Comment:
Likewise here; it would be much nicer to have a standard way of translating the various piston blockType into the cardinal directions, then simply reference that here once as something like `this.retractArm(neighborPosition[direction]);`

Reply:
At the moment, the getOrthogonalPositions just returns an array of 4 2-element-arrays (positions). Thus the numeric indexing. I mean, the """direction""" that controller uses in moveDirection() is the same way (just 0-4 as an int). Which is awful. 

we COULD refactor getOrthogonalPositions to be like getOrthogonalBlocks and return an object that has 4 elements that are other objects called north/south/east/west so that we could reference neighborPosition.north? I don't know how extensive of a refactor that would be though, given that getOrthogonalPositions is used in a few areas?

Let me know what trigger you want to pull on this one.

Reply:
Or I could just #define north 0;

😂 😂 😂 😂

Reply:
Curses! @joshlory beat me to that idea!

const North = 1;
const South = 2;
const East = 4;
const West = 8;

Reply:
This PR has some convenience methods that may help?  https://github.com/code-dot-org/craft/pull/96/files#r138232605

I made a big mistake by adding North/South/East/West.  I'm planning to unify with Up/Down/Right/Left!